File: backend/app/models/task.py
"""
Task Pydantic Models
Based on actual schema: tasks table
"""
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, Field, field_validator


class TaskCreate(BaseModel):
    """Task creation input model"""
    project_id: int = Field(..., ge=1, description="Project ID (required)")
    classification_id: int = Field(..., ge=1, description="Classification ID (required)")
    title: str = Field(..., min_length=1, max_length=500, description="Task title (required)")
    description: Optional[str] = Field(None, max_length=2000, description="Task description")
    status: str = Field(default="open", description="Task status (default: 'open')")
    baseline_start: Optional[datetime] = Field(None, description="Baseline start date/time")
    baseline_end: Optional[datetime] = Field(None, description="Baseline end date/time")
    actual_start_date: Optional[date] = Field(None, description="Actual start date")
    actual_end_date: Optional[date] = Field(None, description="Actual end date")

    @field_validator('baseline_end')
    @classmethod
    def validate_baseline_dates(cls, v, info):
        """Validate baseline_end >= baseline_start"""
        if v and 'baseline_start' in info.data and info.data['baseline_start']:
            if v < info.data['baseline_start']:
                raise ValueError('baseline_end must be >= baseline_start')
        return v

    @field_validator('actual_end_date')
    @classmethod
    def validate_actual_dates(cls, v, info):
        """Validate actual_end_date >= actual_start_date"""
        if v and 'actual_start_date' in info.data and info.data['actual_start_date']:
            if v < info.data['actual_start_date']:
                raise ValueError('actual_end_date must be >= actual_start_date')
        return v


class TaskUpdate(BaseModel):
    """Task update input model (all fields optional)"""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = Field(None, max_length=2000)
    status: Optional[str] = Field(None)
    classification_id: Optional[int] = Field(None, ge=1)
    baseline_start: Optional[datetime] = None
    baseline_end: Optional[datetime] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None

    @field_validator('baseline_end')
    @classmethod
    def validate_baseline_dates(cls, v, info):
        """Validate baseline_end >= baseline_start"""
        if v and 'baseline_start' in info.data and info.data.get('baseline_start'):
            if v < info.data['baseline_start']:
                raise ValueError('baseline_end must be >= baseline_start')
        return v

    @field_validator('actual_end_date')
    @classmethod
    def validate_actual_dates(cls, v, info):
        """Validate actual_end_date >= actual_start_date"""
        if v and 'actual_start_date' in info.data and info.data.get('actual_start_date'):
            if v < info.data['actual_start_date']:
                raise ValueError('actual_end_date must be >= actual_start_date')
        return v


class TaskOut(BaseModel):
    """Task output model"""
    id: int
    project_id: int
    project_name: Optional[str] = None  # JOIN from projects.name
    classification_id: int
    title: str
    description: Optional[str] = None
    status: str
    baseline_start: Optional[datetime] = None
    baseline_end: Optional[datetime] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/types/task.ts
/**
 * Task Types
 * Based on actual DB schema: tasks table
 */
export type TaskStatus = "open" | "closed" | "in_progress" | string;

export interface Task {
  id: number;
  project_id: number;
  project_name: string | null; // JOIN from projects.name
  classification_id: number;
  title: string;
  description: string | null;
  status: TaskStatus;
  baseline_start: string | null; // ISO datetime string
  baseline_end: string | null; // ISO datetime string
  actual_start_date: string | null; // ISO date string
  actual_end_date: string | null; // ISO date string
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
}

export interface TaskCreate {
  project_id: number;
  classification_id: number;
  title: string;
  description?: string | null;
  status?: string;
  baseline_start?: string | null;
  baseline_end?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
}

export interface TaskUpdate {
  title?: string;
  description?: string | null;
  status?: string;
  classification_id?: number;
  baseline_start?: string | null;
  baseline_end?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
}

export interface TaskListParams {
  q?: string; // Search query
  status?: string;
  project_id?: number;
  classification_id?: number;
  limit?: number;
  offset?: number;
  sort?: string;
}

